src/compiler/transformers/es2015.ts

```diff
@@ -1344,6 +1344,38 @@ namespace ts {
             return node && node.dotDotDotToken && node.name.kind === SyntaxKind.Identifier && !inConstructorWithSynthesizedSuper;
         }
 
+        /**
+         * Gets the index of the first statement that contains a reference to a given parameter
+         *
+         * @param restParameter The ParameterDeclaration to which we are finding references.
+         * @param statements An array of Statement nodes to search
+         */
+        function findRestParameterReferenceIndex(restParameter: ParameterDeclaration, statements: Statement[]) {
+            const originalRestParameter = getOriginalNode(restParameter);
+
+            return findIndex(statements, containsReferenceToParameter);
+
+            function containsReferenceToParameter(statement: Statement): boolean {

```

Comment:
This results in a recursive walk of the entire subtree of every statement of a function or method, which we try to avoid in the transforms. It seems like this is best handled in the checker when checking an Identifier. At that point you can look up its declaration, determine whether the declaration is a rest parameter, and walk up from there to the topmost containing statement and set something like a `NodeCheckFlags.ContainsRestParameterReference` on the statement. Walking up the spine is far cheaper than walking down to each descendant.

Note that you would also need to verify here in the transformer that the rest parameter was type checked, as its possible that a transformation added a previously non-existent rest-parameter. In that case, I would err on the side of caution and emit at the top of the function.

Reply:
Thanks, I wasn't aware of trying to avoid recursive walks in transforms. I'll take a look at flagging this info (and the below walk as well) during type checking.

Reply:
While a tree walk is generally cheap, each transform is run in succession resulting in multiple tree walks per file. In the compiler we attempt to limit this as much as possible through various optimizations. One optimization we regularly employ (though it doesn't apply here) are transformation flags that we calculate during our bind step (since bind is a guaranteed top-down full walk of the tree), or check flags that we calculate in the checker (which can randomly access the tree). These flags allow us to avoid full walks of the tree because we can use that information to determine whether a tree walk is necessary for each case.